### DISCUSSION TITLE
createRedirect (redirectInBrowser) to support query parameters

### DISCUSSION THREAD

**Discussion by mustofa-binary** [gatsbyjs/gatsby#27582] [Ideas / Feature Requests] [answered]
## Description

I attempt to create redirect from `/en` to `/`, I am expecting all of the query parameters brought along from the link, will be redirected to the destination link, but it seems to remove all query params or hash.

### Steps to reproduce

Here is my code

```
exports.onCreatePage = ({ page, actions }) => {
    const { createRedirect, createPage, deletePage } = actions

    // First delete the incoming page that was automatically created by Gatsby
    // So everything in src/pages/
    deletePage(page)

    Object.keys(language_config).map((lang) => {
        // Use the values defined in "locales" to construct the path
        const { path, is_default } = language_config[lang]
        const localized_path = is_default ? page.path : `${path}${page.path}`
        const is_production = process.env.GATSBY_ENV === 'production'

        if (is_production) {
            if (path === 'ach') return
        }

        if (!translations_cache[lang]) {
            const translation_json = require(`./src/translations/${lang}`)
            translations_cache[lang] = translation_json
        }

        const current_page = createPage({
            // Pass on everything from the original page
            ...page,
            // Remove trailing slash from page.path (e.g. "/de/")
            path: localized_path,
            // Pass in the locale as context to every page
            context: {
                ...page.context,
                locale: lang,
                localeResources: translations_cache[lang],
                pathname: localized_path,
            },
        })

        if (is_default) {
            const en_path = `/en${localized_path.slice(0, -1)}`
            createRedirect({ fromPath: en_path, toPath: localized_path, redirectInBrowser: true, isPermanent: true})
            createRedirect({ fromPath: `${en_path}/`, toPath: localized_path, redirectInBrowser: true, isPermanent: true})
        }

        return current_page
    })
}
```

- if i put /en/about#story?some_query_params=sdf

### Expected result

it should be redirected along with hash and query params `/about#story?some_query_params=sdf`

### Actual result

- it will become /about . removing all the params and hash

### Environment

Run `gatsby info --clipboard` in your project directory and paste the output here.

  System:
    OS: macOS 10.15.6
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 12.14.1 - /usr/local/bin/node
    Yarn: 1.21.1 - /usr/local/bin/yarn
    npm: 6.14.7 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 86.0.4240.111
    Firefox: 77.0.1
    Safari: 13.1.2
  npmPackages:
    gatsby: ^2.24.63 => 2.24.63 
    gatsby-background-image: ^1.1.2 => 1.1.2 
    gatsby-image: ^2.4.19 => 2.4.19 
    gatsby-plugin-amp: ^0.2.2 => 0.2.2 
    gatsby-plugin-anchor-links: ^1.1.1 => 1.1.1 
    gatsby-plugin-eslint: ^2.0.8 => 2.0.8 
    gatsby-plugin-force-trailing-slashes: ^1.0.4 => 1.0.4 
    gatsby-plugin-manifest: ^2.4.30 => 2.4.30 
    gatsby-plugin-nprogress: ^2.3.11 => 2.3.11 
    gatsby-plugin-react-helmet: ^3.3.11 => 3.3.11 
    gatsby-plugin-react-svg: ^3.0.0 => 3.0.0 
    gatsby-plugin-robots-txt: ^1.5.3 => 1.5.3 
    gatsby-plugin-sharp: ^2.6.36 => 2.6.36 
    gatsby-plugin-sitemap: ^2.4.13 => 2.4.13 
    gatsby-plugin-styled-components: ^3.3.11 => 3.3.11 
    gatsby-plugin-stylelint: ^3.2.0 => 3.2.0 
    gatsby-source-filesystem: ^2.3.30 => 2.3.30 
    gatsby-transformer-sharp: ^2.5.15 => 2.5.15 
  npmGlobalPackages:
    gatsby-cli: 2.12.91
    gatsby: 2.15.15

**Comment by ascorbic** [comment]
Hi,
Yes, I'm afraid this is expected behaviour. I'll change this to a feature request. 
Incidentally, in your example you have the syntax of the url wrong. It should be `/en/about?some_query_params=sdf#story` not `/en/about#story?some_query_params=sdf`. Fragments go at the end of the URL.

**Reply by jonsadka** [reply]
@ascorbic what I'm looking to achieve is something like this:

```js
createRedirect({
  fromPath: 'product?style=abc',
  toPath: 'product-abc'
)}

createRedirect({
  fromPath: 'product?style=xyz',
  toPath: 'product-xyz'
)}
```

two questions:
1) is there a link to the feature request so I can follow?
2) based on your comment, it seems like my desired behavior is not currently possible because `createRedirect` ignores query params?

**Comment by JKlause** [comment]
I am working on a website that uses query params. 
We are also calling `createRedirect` a fair amount in our build process (with AWS s3). 
We'd like these redirects to forward query params (if any are present). 
At the moment that doesn't seem to be the expected behavior of createRedirect (as per this discussion, and the docs).

Is there an update to this feature request request? Any way that we can track its progress?
Thanks for the update!
cc: @ascorbic

**Reply by LekoArts** [reply]
We don't actively work on this. If you have a need for it we're happy to review a PR :)

**Reply by JKlause** [reply]
Ok! So what I can glean, you have no plans of implementing this feature on createRedirect. If I want to, you'll happily look at my PR. And in the short term I should look at AWS as a means of solving the issue?

**Comment by jpvajda** [comment]
We'd love to see this implemented so to solve this problem we are facing.  https://github.com/newrelic/docs-website/issues/2139

**Reply by pragmaticpat** [reply]
@jpvajda @JKlause @jonsadka , et al... 

We have this queued up for our next sprint, which begins on June 28, 2021. **If all goes well**, this behavior should be available in Gatsby by mid-July.

The primary objective of this feature is to ensure that hash and query params are preserved upon redirect, such that the redirectTo page receives the hash and query params the redirectFrom page initially received.

Am I missing some nuance in that outcome?

**Reply by jpvajda** [reply]
@pragmaticpat I checked in with @zstix and  @aswanson-nr  on this question and confirmed that's our expected outcome. Just let us know when this is avaiable in a release and we'll get it resolved on our end. Thank you!

**Reply by jonsadka** [reply]
@pragmaticpat this is our desired outcome as well. Thanks!

**Reply by jpvajda** [reply]
@pragmaticpat 👋  has this been release yet? If not, do you happen to have an ETA or an issue we can track?

**Reply by pragmaticpat** [reply]
@jpvajda - this one is looking like it'll be available in July. When we know the specific Gatsby version number that will contain it, we'll relay back here. Thanks!

**Reply by jpvajda** [reply]
@pragmaticpat thank you!

**Comment by LekoArts** [comment, marked as answer]
We merged https://github.com/gatsbyjs/gatsby/pull/32334 and released it in `gatsby@next` so you could try it out there. It'll then be released on July 20th in 3.10
